### ISSUE TITLE
Sporadical "[Errno 110] Connection timed out" in databricks adapter.

### ISSUE THREAD

**Issue by patraden** [databricks/dbt-databricks#248]
### Describe the bug
We are building dwh with Databricks premium on GCS and decided to use dbt core 1.3.2 (databricks adapter 1.3.2) for ELT pipelines and documentation. So far in production we have developed just 1 model pattern (incremental "merge" updates for staging tables in databricks DeltaLake). ELT pipeline triggers dbt run command with --select option to incrementally update relevant staging tables. However, from time to time, absolutely randomly some runs throws "[Errno 110] Connection timed out" error for random models in selection. This is happening with no pattern, sometimes everything runs fine, sometimes not. Interestingly enough, we also tested this behaviour with Databricks workflows (to check for possible connectivity reasons behind this) and figured out the same behaviour even when dbt is run on databricks dbt cli cluster as per documentation here. What's even more interesting is that problem gets amplified "exponentially" if we use more than 1 thread in profile (now we have to run models sequentially to minimize errors). This problem is becoming a big concern for us about further use of dbt with databricks.. We have also considered "retry_all: true" mechanism described for spark adapter, but it does not seem to be triggered for these types of errors.
```
dbt run --select tag:pg_230_deals --vars '{"env": "prod", "data_interval_start": 1673103600, "data_interval_end": 1673190000}'
....
03:36:13.436533 [info ] [MainThread]: Finished running 14 incremental models in 0 hours 33 minutes and 48.86 seconds (2028.86s).
03:36:13.436860 [debug] [MainThread]: Connection 'master' was properly closed.
03:36:13.436988 [debug] [MainThread]: Connection 'model.databricks.deals__pg_230__tradestone_tp_real_cry_1__mt5_deals' was properly closed.
03:36:13.459647 [info ] [MainThread]: 
03:36:13.460106 [info ] [MainThread]: Completed with 1 error and 0 warnings:
03:36:13.460478 [info ] [MainThread]: 
03:36:13.460738 [error] [MainThread]: Runtime Error in model deals__pg_230__fbs_mt5_real_1__mt5_deals (models/staging/pg_230/deals/deals__pg_230__fbs_mt5_real_1__mt5_deals.sql)
03:36:13.461018 [error] [MainThread]:   Error during request to server: [Errno 110] Connection timed out
03:36:13.461290 [info ] [MainThread]: 
03:36:13.461510 [info ] [MainThread]: Done. PASS=13 WARN=0 ERROR=1 SKIP=0 TOTAL=14
```

### Steps To Reproduce

1. Use multithreading in profiles.yml (threads = 5+)
2. Create several (10+) models
3. Run them selectively (--select)

if required, i can share full github repo for our project.

This is out typical model for staging tables with the same schema:
```
# models.yml
version: 2

models:
  - name: deals__pg_230__fbs_b100_real_1__mt5_deals
    description: "dealing replica pg 230, table: fbs_b100_real_1.mt5_deals"
    config:
      materialized: incremental
      incremental_strategy: merge
      file_format: delta
      on_schema_change: fail
      unique_key: deal
      schema: staging

  - name: deals__pg_230__fbs_levelup_real_2__mt5_deals
    description: "dealing replica pg 230, table: fbs_levelup_real_2.mt5_deals"
    config:
      tags: ['pg_230_deals']
      materialized: incremental
      incremental_strategy: merge
      file_format: delta
      on_schema_change: fail
      unique_key: deal
      schema: staging
...
```

```
# deals__pg_230__fbs_b100_real_1__mt5_deals.sql
{{ config(
        pre_hook="CREATE TABLE if not exists {{ source('dw', 'deals__pg_230__fbs_b100_real_1__mt5_deals') }} USING parquet LOCATION 'gs://staging-databricks-{{ var('env') }}/pg_230__deals/fbs_b100_real_1__mt5_deals/'",
    ) }}
with src as
(
select
 *,
 row_number() over (partition by deal order by timestamp desc) rn
from {{ source('dw', 'deals__pg_230__fbs_b100_real_1__mt5_deals') }}
)
select *
from src
{% if is_incremental() %}
  where rn = 1 and timestamp between '{{ var('data_interval_start') }}' and '{{ var('data_interval_end') }}'
{% endif %}
```
```
# sources.yml
version: 2

sources:
  - name: dw
    description: dw temporary tables or external tables in gcs
    tables:
      - name: deals__pg_230__fbs_b100_real_1__mt5_deals
      - name: deals__pg_230__fbs_levelup_real_1__mt5_deals
 ...
```

### Expected behavior
Expecting more robust and predictable processing of dbt run command with multithreading and clear instruction on how to make operational error handling for intermittent [Errno 110] connectivity errors.

### Screenshots and log output
If applicable, add screenshots or log output to help explain your problem.

### System information
**The output of `dbt --version`:**
```
Core:
  - installed: 1.3.2
  - latest:    1.3.2 - Up to date!

Plugins:
  - spark:      1.3.0 - Up to date!
  - databricks: 1.3.2 - Up to date!
```

**The operating system you're using:**

**The output of `python --version`:**
Python 3.8.10

**Comment by jfox-teine** [comment]
What operating system are you running dbt on? Are you hitting a SQL data warehouse pro as your destination?

I'm running the same using 32 threads against a 'Small' data warehouse with 1-3 clusters, auto scaling on, photon enabled, using this as my dbt cluster config and I can't reproduce this issue:

new_cluster {
      num_workers    = 0
      runtime_engine = "STANDARD"
      spark_version  = data.databricks_spark_version.latest_lts.id
      node_type_id   = var.databricks_default_node_type
      spark_env_vars = {
        PYSPARK_PYTHON = "/databricks/python3/bin/python3"
      }

      spark_conf = {
        # Single-node
        "spark.databricks.cluster.profile" : "singleNode"
        "spark.master" : "local[*, 4]"
      }
      custom_tags = {
        "ResourceClass" = "SingleNode"
      }
    }

**Comment by andrefurlan-db** [comment]
Hi @patraden, thanks for reporting this.
To help us understand, can you run with log level set to debug and provide the dbt.log file?

 Furthermore, can you contact Databricks support so we can look into what is happening on our side?

**Comment by patraden** [comment]
Good day @jfox-teine and @andrefurlan-db ! Thank you for coming back to me! I will try to respond you both in a single message.

1. We are running dbt-core on standard (no extra configs) google cloud compute VM operated by Ubuntu 20.04.5 server (details below)
2. However, as I mentioned, I wanted to make infra agnostic tests and could reproduce the issue in Databricks workflows (databricks cli cluster) as per instructions here.
3. As per your request, I have reproduced the issue (in debug) in Databricks workflows (databricks cli cluster) and collected all the execution artefacts as well as cluster/dbt job configuration in [my GitHub repo](https://github.com/patraden/ddl-dbt-databricks). Thus Databricks job gets initiated from this repo and fails when threads = 32 for all models, but succeeds for all models when threads = 1 (which is a luck, normally 1 or 2 gets failed).
Repo structure is the following:
```
ddl-dbt-databricks/
   databricks/  # dbt project with all the code we have so far
   #248_execution_artefacts/
       databricks_workflows_dbt_job.json  # dbt run job configuration (including dbt cli cluster)
       dbt_cluster_test.json  # main dedicated dbt cluster configuration
       dbt-output-threads-1.tar.gz  # execution artefacts for successful threads = 1 job
       dbt-output-threads-32.tar.gz  # execution artefacts for failed threads = 32 job
```
4. Finally, I want to say that issue might be OS related because when I run jobs locally on my Mac Pro M1 - everything works just fine even with multithreading. I do not know, I am really stuck here on how to get it all setup properly for our google cloud based configuration. I really expected that at least databricks workflows configuration should work "out of the box" but it does not.


```
# https://console.cloud.google.com/compute/instancesDetail
Basic info
OS
Ubuntu 20.04.5 LTS
OS version
20.04
OS Config agent version
20210608.1-0ubuntu1~20.04.1
Patch status
 Updates available
 
# bash
dbt@dbt-databricks:~$ uname -a
Linux dbt-databricks 5.15.0-1025-gcp #32~20.04.2-Ubuntu SMP Tue Nov 29 08:31:04 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jfox-teine** [comment]
Thanks for the detailed info @patraden. For clarification, I'm just another customer curious about this issue. For testing purposes I can't 100% recreate as I'm not using the Google platform. I can however recreate a test in databricks using the Azure platform using my own models. Using a Standard_D4ds_v5 single node worker I'm able to execute 32 threads. Here's my cluster setup for the dbt task:

```
{
    "num_workers": 0,
    "cluster_name": "",
    "spark_version": "11.3.x-scala2.12",
    "spark_conf": {
        "spark.databricks.delta.preview.enabled": "true",
        "spark.databricks.cluster.profile": "singleNode",
        "spark.master": "local[*, 4]"
    },
    "azure_attributes": {
        "first_on_demand": 1,
        "availability": "ON_DEMAND_AZURE",
        "spot_bid_max_price": -1
    },
    "node_type_id": "Standard_D4ds_v5",
    "ssh_public_keys": [],
    "custom_tags": {
        "ResourceClass": "SingleNode"
    },
    "spark_env_vars": {
        "PYSPARK_PYTHON": "/databricks/python3/bin/python3"
    },
    "enable_elastic_disk": true,
    "cluster_source": "JOB",
    "init_scripts": [],
    "data_security_mode": "SINGLE_USER",
    "runtime_engine": "STANDARD"
}
```

**Comment by andrefurlan-db** [comment]
Hi @patraden , it could indeed be OS related as socket timeout's vary across OS. Can you try to set it up to a high number to see if it works? Just add this to your profiles.yml.
yaml
```
    dev:
      type: databricks
      ....
      connection_parameters:  
        _socket_timeout: 60000
```

**Comment by patraden** [comment]
Hi @andrefurlan-db I tried that today, but unfortunately does not seem it made any effect. Errors continue

**Comment by andrefurlan-db** [comment]
Hi @patraden, strange. I thought this could be the same root cause as https://github.com/databricks/databricks-sql-python/pull/24. Did you manage to contact Databricks support? I'll need to extra access to check if we have any logs on our side with any clue of what is going on.
Also, just to narrow down things a bit, are you able to reproduce the problem by only using https://pypi.org/project/databricks-sql-connector/ ? I can't with the clusters/warehouses I have available for me. I don't think this is a problem with the dbt-adapter (but I might be wrong).

**Comment by patraden** [comment]
HI @andrefurlan-db! Sorry for not getting back to you in a while. Had few busy days... I have some news though on the issue... I dedicated some time for debugging and decided to start with OS assumptions. Thus I have rebuilt previous google cloud vm operated by Ubuntu to RHEL and all of a sudden problem have gone! I was able to run models concurrently with multithreading and  parallel ssh sessions. No issues at all! I do not know, weather this is google ubuntu image or ubuntu itself but this is definitely linked to Ubuntu alone (or maybe all debians).
